### PR TITLE
Prepare for model training, intermediate PR.

### PR DESCRIPTION
**Pull request by SeonghwanSeo**
### Finish
- Test structure sampling (forward is working well; but it should be checked that there is no logic error)
- Add script for weighted pdb sampling
- Update data structure for token-level cropping
  - original: `BoltzStructure` -> `FoldingInput`
  - now: `BoltzStructure` -> `TokenizedStructure` -> `FoldingInput`

### TODO
- Add cropping algorithm (contiguous;spatial;interface-spatial)
- Add DataModule
- Finish TrainingModule
  - add optimizer and lr scheduler
  - validation: TODO (next PR)

### CODE REVIEW

File: src/kfold/training/folding/loss/distogram.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        # Get mask
+        mask = f_input.token.disto_mask  # [B, Lt]
+        pair_mask = mask[:, None, :] * mask[:, :, None]  # [B, Lt, Lt]
+        pair_mask.diagonal(dim1=-2, dim2=-1).zero_()  # zero out diagonal
+        pair_mask = pair_mask.to(dtype=logits.dtype)
+
+        # Compute the distogram loss
+        disto_loss = torch.nn.functional.cross_entropy(
+            logits.view(B * L * L, self.num_bins),
+            target_distogram.view(),
```
[the remark this answers is not in the document]
Comment (SeonghwanSeo) [reply src/kfold/training/folding/loss/distogram.py]: update.

File: src/kfold/utils/boltz/process.py [hunk 2]
```diff
@@ -203,262 +226,181 @@ def get_dtype(key: str) -> torch.dtype:
...
-        asym_id=chain_info["asym_id"].view(Nc),
-        sym_id=chain_info["sym_id"].view(Nc),
-        num_residues=chain_info["num_residues"].view(Nc),
-        num_atoms=chain_info["num_atoms"].view(Nc),
-        num_tokens=chain_info["num_tokens"].view(Nc),
-        pad_mask=chain_info["pad_mask"].view(Nc),
+    # TODO: use actual apo coords when available
+    atom_arr["apo_coords"] = np.zeros_like(atom_arr["label_coords"])
+    for num_tokens in chain_arr["num_tokens"]:
+        start_idx = np.sum(chain_arr["num_tokens"][:num_tokens])
```
[the remark this answers is not in the document]
Comment (SeonghwanSeo) [reply src/kfold/utils/boltz/process.py:353]: `num_tokens` is array.

File: src/kfold/model/modules/structure_module/af3_edm.py [hunk 3]
```diff
@@ -320,17 +319,12 @@ def sample_holo(
...
 
-        B, N, L = holo_coords.shape[:3]
-        holo_coords = holo_coords.view(B * N, L, 3)  # (B*N, Latom, 3)
-        atom_mask = atom_mask.repeat_interleave(N, dim=0)  # (B * N, Latom)
-
         # Apply coordinate augmentation
         holo_coords = self.random_augmentation(holo_coords, atom_mask=atom_mask)
 
         # Mask out the padding atoms
         holo_coords = holo_coords * atom_mask[:, :, None]  # (B*N, Latom, 3)
```
[the remark this answers is not in the document]
Comment (SeonghwanSeo) [reply src/kfold/model/modules/structure_module/af3_edm.py]: I'll update.

File: src/kfold/training/folding/training_module.py [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+
+    @ema.setter
+    def ema(self, value: ExponentialMovingAverage):
+        self._ema = value
+
+    def on_train_start(self):
+        if self.use_ema:
+            if not self.is_ema_initialized:
+                self.ema = ExponentialMovingAverage(
+                    parameters=self.parameters(), decay=self.ema_decay
```
[the remark this answers is not in the document]
Comment (SeonghwanSeo) [reply src/kfold/training/folding/training_module.py]: I'll update.

File: src/kfold/model/modules/trunk/af3_trunk.py [hunk 5]
```diff
@@ -158,12 +158,12 @@ def forward(
         # Line 6, z_hat, s_hat = 0, 0
         z_hat, s_hat = z_init, s_init  # just to make sure the types are correct
 
-        for i in range(num_recycles + 1):
-            no_grad = self.training and (i < num_recycles)
+        for i in range(1, num_cycles + 1):
+            no_grad = self.training and (i == num_cycles)
```
[the remark this answers is not in the document]
Comment (SeonghwanSeo) [reply src/kfold/model/modules/trunk/af3_trunk.py]: I accept.

File: src/kfold/data/tokenized.py [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+            asym_id = cropped_chain.asym_id[cidx]
+            token_mask = cropped_token.asym_id == asym_id
+            num_tokens = np.sum(token_mask).item()
+            num_residues = np.sum(
+                np.unique(cropped_token.residue_index[token_mask])
+            ).item()
+            num_atoms = np.sum(cropped_token.num_atoms[token_mask]).item()
+            cropped_chain.num_tokens[cidx] = num_tokens
+            cropped_chain.num_residues[cidx] = num_residues
+            cropped_chain.num_atoms[cidx] = num_atoms
```
[the remark this answers is not in the document]
Comment (SeonghwanSeo) [reply src/kfold/data/tokenized.py:347]: I think this is fine. I already copy. numpy in-space operation.

File: src/kfold/model/models/kfold.py [hunk 7]
```diff
@@ -51,44 +51,192 @@ def __init__(self, global_config: DictConfig):
...
+                z_trunk,
+                diffusion_batch_size,
+                model_cache=model_cache,
+            )
+
+        if train_confidence_module:
+            # TODO: implement confidence prediction with mini-rollout
+            coordinates = dict_out["sample"]["coordinates"]
+            s_trunk_detached = s_trunk.detach()  # noqa
+            z_trunk_detached = z_trunk.detach()  # noqa
```
[the remark this answers is not in the document]
Comment (SeonghwanSeo) [reply src/kfold/model/models/kfold.py:210]: This is for todo.

File: src/kfold/model/models/kfold.py [hunk 8]
```diff
@@ -51,44 +51,192 @@ def __init__(self, global_config: DictConfig):
...
+                s_inputs,
+                s_trunk,
+                z_trunk,
+                diffusion_batch_size,
+                model_cache=model_cache,
+            )
+
+        if train_confidence_module:
+            # TODO: implement confidence prediction with mini-rollout
+            coordinates = dict_out["sample"]["coordinates"]
```
[the remark this answers is not in the document]
Comment (SeonghwanSeo) [reply src/kfold/model/models/kfold.py:208]: This is for todo.

File: src/kfold/model/models/kfold.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (SeonghwanSeo) [reply src/kfold/model/models/kfold.py:210]: This is for todo.

File: src/kfold/data/layout.py [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+    @property
+    def length(self) -> int:
+        """Get the length of the layout."""
+        return self.layout_shape[0]
+
+    def __len__(self) -> int:
+        """The length of the layout."""
+        return self.length
+
+    def __getitem__(self, idx: int | slice | ArrayT) -> Self:
```
[the remark this answers is not in the document]
Comment (SeonghwanSeo) [reply src/kfold/data/layout.py:125]: That is fine.

File: tests/test_process.py [hunk 10]
```diff
@@ -197,41 +280,38 @@ def check_structure(key: str, verbose: bool = False):
...
-
-        keys = keys[:10000]
+    manifest = {v["id"]: v for v in manifest}
+    keys = sorted(list(manifest.keys()))
+    random.seed(42)
+    random.shuffle(keys)
 
+    keys = keys[:10000]
+    if False:
+        """Multi-processing test"""
```
[the remark this answers is not in the document]
Comment (SeonghwanSeo) [reply tests/test_process.py:314]: This is for debugging purpsoe.
